python/tank/platform/environment.py

```diff
@@ -329,8 +328,24 @@ def get_app_descriptor(self, engine_name, app_name):
 
         return d
 
-    def set_context(self, context):
-        self.__context = context
+    def change_context(self, new_context):
+        """
+        Sets the environment's context and refreshes all environment data.
+
+        :param new_context: The new sgtk.context.Context to switch to.
+        """
+        self.__context = new_context
+
+        # We need to figure out the env file path before we can
+        # refresh. We run it through the new context through the
+        # core hook, which will give us the name (like shot_step),
+        # and then get the filesystem path to that .yml file.
+        sgtk = new_context.tank
+        env_name = sgtk.execute_core_hook(
+            constants.PICK_ENVIRONMENT_CORE_HOOK_NAME,
+            context=new_context,
+        )
+        self._env_path = sgtk.pipeline_configuration.get_environment_path(env_name)
```

Comment:
This seems scary. Suddenly the whole internal representation of the environment object can change underneath us. 

The intent with the environment object is that it represents an environment and allows you to retrieve  properties etc. The `WritableEnvironment` represents an mutable version of this where you can go ahead and perform updates etc on the environment. But fundamentally the `Environment` wraps around single env file and this never changes.

The whole use context object in the environment class isn't great already as it is - it was added in order for the include system to be able to resolve context specific includes, which are rarely used. I wonder if we could either refactor this so that context is not a member variable but passed in as a param to certain methods or we could consider getting rid of the dependency altogether. Would be nice to brainstorm this a bit! Curious what you think.